Kubelet cluster-wide root CA (bsc#1155810) (#602)
SUSE/doc-caasp: adoc/admin-security-certificates.adoc

During the installation of {productname}, a CA (Certificate Authority) certificate is generated,
which is then used to authenticate and verify all communication. This process also creates
and distributes client and server certificates for the components.

== Communication Security
Communication is secured with TLS v1.2 using the AES 128 CBC cipher.
All certificates are 2048 bit RSA encrypted.

== Certificate Validity
The CA certificate is valid for 3650 days (10 years) by default.
Client and server certificates are valid for 365 days (1 year) by default.

== Certificate Location
Required CAs for {productname} are stored on all master nodes:

[%header,cols=3*]
|===
|Common Name
|Path
|Description

|kubernetes
|/etc/kubernetes/pki/ca.crt,key
|kubernetes general CA

|etcd-ca
|/etc/kubernetes/pki/etcd/ca.crt,key
|Etcd cluster

|kubelet-ca
|/var/lib/kubelet/pki/ca.crt,key

> |Kubelet components

Reviewer: For me it's not ca.crt,key but kubelet-ca.crt,key.

Am I wrong ?

Author: Thanks, my fault, typo.

Author: #620

|Client
|===

== Deployment with a Custom CA Certificate

[WARNING]
====
Please plan carefully when deploying with a custom CA certificate. This certificate
can not be reconfigured once deployed and requires a full re-installation of the
cluster to replace.
====

Administrators can provide custom CA certificates (root CAs or intermediate CAs) 

> during cluster deployment and decide which CA components to replace (multiple CA certificates) or if to replace all as one (single CA certificate).

Reviewer 2: ```suggestion
during cluster deployment and decide which CA components to replace (multiple CA certificates) or if to replace all with a single CA certificate.
```

. Log in to the master node and regenerate the certificates:
+
* Replace the oidc-dex server certificate:
+
Sign the oidc-dex server certificate with the CA `ca.crt/ca.key` and SAN

> as origin was. Then update the {kube} cluster secret data

Reviewer 2: I don't quite understand this "as origin was" bit. :) What is origin in this case?

`ca.crt`, `tls.crt`, and `tls.key` with base64 encoded:
+
[source,bash]
----
sudo mkdir -p /etc/kubernetes/pki.bak
cd /etc/kubernetes/pki.bak
sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf get secret oidc-dex-cert -n kube-system -o yaml | sudo tee oidc-dex-cert.yaml > /dev/null

sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf edit secret oidc-dex-cert -n kube-system
sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf delete pod -lapp=oidc-dex -n kube-system
----

* Replace the oidc-gangway server certificate:
+
Sign the oidc-gangway server certificate with the CA `ca.crt/ca.key` and SAN

> as origin was. Then update the {kube} cluster secret data

Reviewer 2: I don't quite understand this "as origin was"  bit. :) What is origin in this case?

Author: @nkoranova I updated the detailed steps on how to get origin SAN field. Hope this is more clear for you :)

Reviewer 2: Thanks! I've suggested changes. :)

`ca.crt`, `tls.crt`, and `tls.key` with base64 encoded.
+
[source,bash]
----
sudo mkdir -p /etc/kubernetes/pki.bak
cd /etc/kubernetes/pki.bak
sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf get secret oidc-gangway-cert -n kube-system -o yaml | sudo tee oidc-gangway-cert.yaml > /dev/null

sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf edit secret oidc-gangway-cert -n kube-system
sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf delete pod -lapp=oidc-gangway -n kube-system
----

* Replace the kubelet server certificate:
+
[IMPORTANT]
====

> You need to generate kubelet server certificate for all the nodes, and the kubelet server key only exists on the control plane nodes. Therefore, after re-signed kubelet server certificate/key for worker nodes, you have to copy the worker nodes kubelet server certificate/key from control plane node.

Reviewer 3: ```suggestion
You need to generate kubelet server certificate for all the nodes on one of control plane nodes, because the kubelet CA certificate key only exists on the control plane nodes. Therefore, after generated re-signed kubelet server certificate/key for worker nodes, you have to copy each kubelet server certificate/key from control plane node to the corresponding worker node.
```

Author: Appreciated.

====
+

> Sign each node kubelet server certificate with the CA `kubelet-ca.crt/kubelet-ca.key` and SAN as origin was. Then update the kubelet server certificate and key file

Reviewer 2: See comment above.

. Log in to the master node and regenerate the certificates:
+
* Replace the oidc-dex server certificate:
+
Backup the original oidc-dex server certificate and key from Secret resource.
+
[source,bash]
----
sudo mkdir -p /etc/kubernetes/pki.bak
sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf get secret oidc-dex-cert -n kube-system -o yaml | sudo tee /etc/kubernetes/pki.bak/oidc-dex-cert.yaml > /dev/null

cat /etc/kubernetes/pki.bak/oidc-dex-cert.yaml | grep tls.crt | awk '{print $2}' | base64 --decode | sudo tee /etc/kubernetes/pki.bak/oidc-dex.crt > /dev/null
cat /etc/kubernetes/pki.bak/oidc-dex-cert.yaml | grep tls.key | awk '{print $2}' | base64 --decode | sudo tee /etc/kubernetes/pki.bak/oidc-dex.key > /dev/null
----
+

> Sign the oidc-dex server certificate with the CA certificate/key `/etc/kubernetes/pki/ca.crt` and `/etc/kubernetes/pki/ca.key`, make sure that the signed server certificate SAN have to be the same as origin is. To get the origin SAN IP address(es) and DNS(s):

Reviewer 2: ```suggestion
Sign the oidc-dex server certificate with the CA certificate/key `/etc/kubernetes/pki/ca.crt` and `/etc/kubernetes/pki/ca.key`, make sure that the signed server certificate SAN is the same as the origin. To get the original SAN IP address(es) and DNS(s) run:
```

+
[source,bash]
----
cat <SIGNED_OIDC_DEX_SERVER_CERT_PATH> | base64 | awk '{print}' ORS='' && echo
cat <SIGNED_OIDC_DEX_SERVER_KEY_PATH> | base64 | awk '{print}' ORS='' && echo

sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf edit secret oidc-dex-cert -n kube-system
sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf delete pod -lapp=oidc-dex -n kube-system
----

* Replace the oidc-gangway server certificate:
+
Backup the original oidc-gangway server certificate and key from Secret resource.
+
[source,bash]
----
sudo mkdir -p /etc/kubernetes/pki.bak
sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf get secret oidc-gangway-cert -n kube-system -o yaml | sudo tee /etc/kubernetes/pki.bak/oidc-gangway-cert.yaml > /dev/null

cat /etc/kubernetes/pki.bak/oidc-gangway-cert.yaml | grep tls.crt | awk '{print $2}' | base64 --decode | sudo tee /etc/kubernetes/pki.bak/oidc-gangway.crt > /dev/null
cat /etc/kubernetes/pki.bak/oidc-gangway-cert.yaml | grep tls.key | awk '{print $2}' | base64 --decode | sudo tee /etc/kubernetes/pki.bak/oidc-gangway.key > /dev/null
----
+

> Sign the oidc-gangway server certificate with the CA certificate/key `/etc/kubernetes/pki/ca.crt` and `/etc/kubernetes/pki/ca.key`, make sure that the signed server certificate SAN have to be the same as origin is. To get the origin SAN IP address(es) and DNS(s):

Reviewer 2: ```suggestion
Sign the oidc-gangway server certificate with the CA certificate/key `/etc/kubernetes/pki/ca.crt` and `/etc/kubernetes/pki/ca.key`, make sure that the signed server certificate SAN is the same as the origin. To get the original SAN IP address(es) and DNS(s), run:
```

----
+
Finally, update the {kube} cluster secret data `tls.crt`, and `tls.key` with base64 encoded from signed oidc-gangway server certificate and key respectively, and restart oidc-gangway pods.
+
[source,bash]
----
cat <SIGNED_OIDC_GANGWAY_SERVER_CERT_PATH> | base64 | awk '{print}' ORS='' && echo
cat <SIGNED_OIDC_GANGWAY_SERVER_KEY_PATH> | base64 | awk '{print}' ORS='' && echo

sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf edit secret oidc-gangway-cert -n kube-system
sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf delete pod -lapp=oidc-gangway -n kube-system
----

* Replace the kubelet server certificate:
+
[IMPORTANT]
====
You need to generate kubelet server certificate for all the nodes on one of control plane nodes, because the kubelet CA certificate key only exists on the control plane nodes. Therefore, after generating re-signed kubelet server certificate/key for worker nodes, you have to copy each kubelet server certificate/key from the control plane node to the corresponding worker node.
====
Backup the original kubelet certificates and keys.
+
[source,bash]
----
sudo cp -r /var/lib/kubelet/pki /var/lib/kubelet/pki.bak
----
+

> Sign each node kubelet server certificate with the CA certificate/key `/var/lib/kubelet/pki/kubelet-ca.crt` and `/var/lib/kubelet/pki/kubelet-ca.key`, make sure that the signed server certificate SAN have to be the same as origin is. To get the origin SAN IP address(es) and DNS(s):

Reviewer 2: ```suggestion
Sign each node kubelet server certificate with the CA certificate/key `/var/lib/kubelet/pki/kubelet-ca.crt` and `/var/lib/kubelet/pki/kubelet-ca.key`, make sure that the signed server certificate SAN is the same as the origin. To get the original SAN IP address(es) and DNS(s) run:
```
